src/main/java/com/dezorganizacja/bridge/view/graphical/FXApplication.java

```diff
@@ -16,14 +23,60 @@ public static void setMainPresenter(MainPresenter mainPresenter) {
 
     @Override
     public void start(Stage stage) throws Exception {
-        FXMLLoader loader = new FXMLLoader(getClass().getResource("/fxml/main.fxml"));
-        Parent root = loader.load();
+        FXApplication.stage = stage;
+        FXApplication.mainPresenter.addObserver(this);
+        loadMainScene();
+    }
+
+    private void loadMainScene() {
+        String programState = mainPresenter.getState();
+        String fxml = stateToFXML(programState);
+
+        if (FXApplication.fxml != null && FXApplication.fxml.equals(fxml)) {
+            return;
+        }
+
+        FXMLLoader loader = new FXMLLoader(getClass().getResource(fxml));
+        Parent root;
+        try {
+            root = loader.load();
+        } catch (IOException e) {
+            System.err.println("error while loading " + fxml + " resource");
+            e.printStackTrace();
+            return;
+        }
+
         stage.setTitle("Bridge");
         stage.setScene(new Scene(root, 300, 275));
 
-        MainController controller = loader.getController();
-        controller.init(mainPresenter);
+        Controller controller = loader.getController();
+        if (controller != null) {
+            controller.init(mainPresenter);
+        }
 
         stage.show();
     }
+
+    public void changeMainScene() {
+        // TODO thread safety
```

Comment:
Jest ok. to jest kolejka FIFO więc rządasz by na coś zmienić: leci na kolejkę. I sie na to zmienia, jesli przed zmianą użytkownik kliknął na coś innego no to i tak sie zmieni na to coś innego w końcu. Dodatkowo masz w loadMainScene sprawdzenie czy nie zmieniasz przypadkiem na to samo wiec wydaje mi się, że wszystko jest ok. Jeśli jest jakiś ciąg wydarzeń w którym coś się może zepsuć z chęcią go usłyszę.